src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorSupervisor.scala

```diff
@@ -176,46 +177,41 @@ class ClusterMonitorSupervisor(monitorConfig: MonitorConfig, dataprocConfig: Dat
-                      }
-                      false // max kernel active time has not been exceeded
-                    }
-                  }
-                  if (maxKernelActiveTimeExceeded){
-                    Metrics.newRelic.incrementCounterIO("maxKernelExceededClusters").unsafeRunAsync(_ => ())
-                    logger.info(s"Auto pausing ${cluster.googleProject}/${cluster.clusterName} due to exceeded max kernel active time")
-                  } else {
-                    Metrics.newRelic.incrementCounterIO("activeKernelClusters").unsafeRunAsync(_ => ())
-                    logger.info(s"Not going to auto pause cluster ${cluster.googleProject}/${cluster.clusterName} due to active kernels")
-                  }
-                  maxKernelActiveTimeExceeded
-
-                } else isIdle
+    pauseableClusters <- clusters.toList.filterA[Future] { cluster =>
+      jupyterProxyDAO.isAllKernalsIdle(cluster.googleProject, cluster.clusterName).attempt.map {
+        case Left(t) =>
+          logger.error(s"Fail to get kernel status for ${cluster.googleProject}/${cluster.clusterName} due to $t")
+          true
+        case Right(isIdle) =>
+          if (!isIdle) {
+            val idleLimit = Duration.ofNanos(autoFreezeConfig.maxKernelBusyLimit.toNanos) // convert from FiniteDuration to java Duration
+            val maxKernelActiveTimeExceeded = cluster.auditInfo.kernelFoundBusyDate match {
+              case Some(attemptedDate) => Duration.between(attemptedDate, Instant.now()).compareTo(idleLimit) == 1
+              case None => {
+                dbRef.inTransaction { dataAccess =>
+                  dataAccess.clusterQuery.updateKernelFoundBusyDate(cluster.id, Instant.now())
+                }
+                false // max kernel active time has not been exceeded
+              }
             }
+            if (maxKernelActiveTimeExceeded){
+              Metrics.newRelic.incrementCounterIO("maxKernelExceededClusters").unsafeRunAsync(_ => ())
+              logger.info(s"Auto pausing ${cluster.googleProject}/${cluster.clusterName} due to exceeded max kernel active time")
+            } else {
+              Metrics.newRelic.incrementCounterIO("activeKernelClusters").unsafeRunAsync(_ => ())
+              logger.info(s"Not going to auto pause cluster ${cluster.googleProject}/${cluster.clusterName} due to active kernels")
+            }
+            maxKernelActiveTimeExceeded
+
+          } else isIdle
         }
     }
     _ <- Metrics.newRelic.gauge("autopausedClusters", pauseableClusters.length).unsafeToFuture()
-    _ <- pauseableClusters.traverse{
-      cl =>
-        logger.info(s"Auto freezing cluster ${cl.clusterName} in project ${cl.googleProject}")
-        leonardoService.internalStopCluster(cl).attempt.map { e =>
-          e.fold(t => logger.warn(s"Error occurred auto freezing cluster ${cl.projectNameString}", e), identity)
-        }
+    _ <- pauseableClusters.traverse[Future, Unit] { cl =>
```

Comment:
Same here -- makes Intellij not complain.

Reply:
new IJ should be happy without the explicit types